### PR TITLE
Adapt components for side chain

### PR DESCRIPTION
**Pull request by SebastianMarian**
## Description of the reasoning behind the pull request (what feature was missing / how the problem was manifesting itself / what was the motive behind the refactoring)

* Adapted BlockTracker and ForkDetector component for side chain approach
- 
- 
  
## Proposed Changes
- 
- 
- 

## Testing procedure
- 
- 
-

### CODE REVIEW

**Review by sasurobert — DISMISSED** [review]

File: process/sync/sideChainShardForkDetector.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+	if shardForkDetector == nil {
+		return nil, process.ErrNilForkDetector
+	}
+
+	return &sideChainShardForkDetector{
+		shardForkDetector,
+	}, nil
+}
+
+func (scsfd *sideChainShardForkDetector) doJobOnBHProcessed(
```
Comment (iulianpascalau) [inline process/sync/sideChainShardForkDetector.go:23]: this does not work as expected

File: process/track/sideChainBlockProcessor.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+	if blockProcessor == nil {
+		return nil, process.ErrNilBlockProcessor
+	}
+
+	return &sideChainBlockProcessor{
+		blockProcessor,
+	}, nil
+}
+
+func (scbp *sideChainBlockProcessor) shouldProcessReceivedHeader(headerHandler data.HeaderHandler) bool {
```
Comment (iulianpascalau) [inline process/track/sideChainBlockProcessor.go:23]: also here

File: process/track/sideChainShardBlockTrack.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	sideChainBlockTrack.blockProcessor = newBlockProcessor
+	return sideChainBlockTrack, nil
+}
+
+// ComputeLongestSelfChain computes the longest chain from self shard
+func (scsbt *sideChainShardBlockTrack) ComputeLongestSelfChain() (data.HeaderHandler, []byte, []data.HeaderHandler, [][]byte) {
```
Comment (iulianpascalau) [inline process/track/sideChainShardBlockTrack.go:37]: this will work as expected if it is only called from exterior. Also the next one

**Review by iulianpascalau — CHANGES_REQUESTED** [review]

File: process/sync/sideChainShardForkDetector.go (hunk 1 above)
Comment (SebastianMarian) [reply process/sync/sideChainShardForkDetector.go:26]: Fixed

File: process/track/sideChainBlockProcessor.go (hunk 2 above)
Comment (SebastianMarian) [reply process/track/sideChainBlockProcessor.go:27]: Fixed

File: process/track/sideChainShardBlockTrack.go (hunk 3 above)
Comment (SebastianMarian) [reply process/track/sideChainShardBlockTrack.go:40]: Fixed

File: process/sync/metaForkDetector.go [hunk 4]
```diff
@@ -15,6 +15,7 @@ var _ process.ForkDetector = (*metaForkDetector)(nil)
 // metaForkDetector implements the meta fork detector mechanism
 type metaForkDetector struct {
 	*baseForkDetector
+	doJobOnBHProcessedFunc func(_ data.HeaderHandler, _ []byte, _ []data.HeaderHandler, _ [][]byte)
```
Comment (iulianpascalau) [inline process/sync/metaForkDetector.go:18]: could have kept the parameters names as it provides better context

**Review by sasurobert — DISMISSED** [review]

File: process/sync/shardForkDetector.go [hunk 5]
```diff
@@ -16,6 +16,7 @@ var _ process.ForkDetector = (*shardForkDetector)(nil)
 // shardForkDetector implements the shard fork detector mechanism
 type shardForkDetector struct {
 	*baseForkDetector
+	doJobOnBHProcessedFunc func(_ data.HeaderHandler, _ []byte, _ []data.HeaderHandler, _ [][]byte)
```
Comment (iulianpascalau) [inline process/sync/shardForkDetector.go:19]: also here, could have kept the parameter names

File: process/track/baseBlockTrack.go [hunk 6]
```diff
@@ -51,6 +51,8 @@ type baseBlockTrack struct {
 	mutHeaders                  sync.RWMutex
 	headers                     map[uint32]map[uint64][]*HeaderInfo
 	maxNumHeadersToKeepPerShard int
+	ComputeLongestSelfChainFunc func() (data.HeaderHandler, []byte, []data.HeaderHandler, [][]byte)
```
Comment (iulianpascalau) [inline process/track/baseBlockTrack.go:54]: Let's not use these exported function pointers. Reason: I can take the component, set this function pointer to nil and then the component will break in panic

File: process/track/blockProcessor.go [hunk 7]
```diff
@@ -25,7 +25,8 @@ type blockProcessor struct {
 	finalMetachainHeadersNotifier         blockNotifierHandler
 	roundHandler                          process.RoundHandler
 
-	blockFinality uint64
+	blockFinality                   uint64
+	shouldProcessReceivedHeaderFunc func(_ data.HeaderHandler) bool
```
Comment (iulianpascalau) [inline process/track/blockProcessor.go:29]: Also here, could have kept the original param names

File: process/track/baseBlockTrack.go (hunk 6 above)
Comment (SebastianMarian) [reply process/track/baseBlockTrack.go:54]: Removed, as it is used only from outside of the component

File: process/track/blockProcessor.go (hunk 7 above)
Comment (SebastianMarian) [reply process/track/blockProcessor.go:29]: Removed, as it is used only from outside of the component and is not necessary to be overwritten

Comment (SebastianMarian) [reply process/track/blockProcessor.go:29]: Done

File: process/sync/shardForkDetector.go (hunk 5 above)
Comment (SebastianMarian) [reply process/sync/shardForkDetector.go:19]: Done

**Review by iulianpascalau — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Pull request merged**